clang/lib/Sema/SemaExpr.cpp

```diff
@@ -6375,12 +6383,16 @@ ExprResult Sema::BuildCXXDefaultInitExpr(SourceLocation Loc, FieldDecl *Field) {
   ImmediateCallVisitor V(getASTContext());
   if (!NestedDefaultChecking)
     V.TraverseDecl(Field);
-  if (V.HasImmediateCalls) {
+  if (V.HasImmediateCalls || InLifetimeExtendingContext) {
     ExprEvalContexts.back().DelayedDefaultInitializationContext = {Loc, Field,
                                                                    CurContext};
     ExprEvalContexts.back().IsCurrentlyCheckingDefaultArgumentOrInitializer =
         NestedDefaultChecking;
 
+    // Pass down lifetime extending flag, and collect temporaries in
+    // CreateMaterializeTemporaryExpr when we rewrite the call argument.
+    keepInLifetimeExtendingContext();
+    keepInMaterializeTemporaryObjectContext();
```

Comment:
Where are the tests for BuildCXXDefaultInitExpr?

Reply:
Thanks for your review! I still have some questions about member init.
For this case(https://godbolt.org/z/jPKe8M6xE), should we materialize the temporaries in class Z::x's default init,  but unless I'm mistaken?

Reply:
@cor3ntin I tried this code https://godbolt.org/z/n77v9rMTz , and the temporaries which in `CXXDefaultInitExpr`  has been lifetime extended in AST. But the LLVM IR unexpected, and the lifetime of the temporaries was not extended. Could you please guide me as to why this is and whether I have made some mistake?

Reply:
I think part of the problem might be that we should only extend in the aggregate case.
ie, when we call `BuildCXXDefaultInit` from `FillInEmptyInitForField`.

In the constructor case, the init lifetime is bound to that of the member initializer expression.

Reply:
Thanks for your help! I have a case use aggregate initialization (https://godbolt.org/z/vn8Y11ja8). I think neither the brace-or-equals initializer nor default member initializer should not be materialized and lifetime-extended in this code snippet, seems they don't satisfy any of [class.temporary P2](http://eel.is/c++draft/class.temporary#2) . @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended? I saw you comment in D153701 that we need to do lifetime-extend for `CXXDefaultInitExpr`. I'm sorry that I'm unfamiliar with that part of clang and I have no clue how to write tests for CXXDefaultInitExpr.

Reply:
I think you are right. 
I'm sorry, that whole thing is a bit confusing to me.

But the initializer is used in the construction of a subobject, which is different from a default argument which is initialized on call site... so I think the conclusion is that we should do nothing for default inits, I think

Reply:
Thank you for your consideration.  I think so too, whether I have made some mistake, I will try to ask the author of P2718R0

Reply:
> @hubert-reinterpretcast Could you please give me an example where `CXXDefaultInitExpr` has to be lifetime-extended?

https://github.com/llvm/llvm-project/issues/85613

Reply:
Many thanks for your information.

Reply:
I try to extend lifetime of the `MaterializedTemporaryExpr`s in `CXXDefaultInitExpr`, but looking from the generated IR, `MaterializedTemporaryExpr ::setExtendingDecl` does not take effect. I think some modifications may be needed in CodeGen.